### ISSUE TITLE
[RFC] Dynamic theme settings + global settings variable

### ISSUE THREAD

**Issue by steffenbrem** [Sylius/Sylius#3985]
Hi,

For my project I was in need of a theme system that is very dynamic, my theme configuration lives inside a database. I am curious in to what extend the Sylius theme bundle meets my approach. I am parsing the following YAML file (the theme schema):

```yaml
name: Dark Theme
version: 1.0
settings:
    -   label: Frontpage product collection
        settings:
            home_featured_products:
                type: taxon
                label: Frontpage product collection
            home_featured_show_header:
                type: boolean
                label: Show header of taxon?
```

Now in the template files of that theme you have a global variable `settings`, which you can use to fetch configuration values. In the administration panel of my e-com. project, the site owner can assign values to the theme configurations like this:
![Screenshot](http://puu.sh/mKrGl/49f24cb5c3.png)

You can get a reference to the selected taxon, by using `{{ settings.home_featured_products }}`. A good example implementation for the theme configuration above is something like this:

```twig
<h3>{{settings.home_featured_products.name }}</h3>
<div class="row collection">
    {% for product in settings.home_featured_products.products(4) %}
        <div class="col-xs-3">
            <div class="collection-item">
                <a href="{{ path(product) }}">
                    <h4>{{ product.name }}</h4>
                </a>
            </div>
        </div>
    {% endfor %}
</div>
```

This way you can build up your theme with all kinds of configurations, only used for that specific theme. I am curious if this is the path Sylius is going or wants to go?

**Comment by pjedrzejewski** [comment]
@steffenbrem Definitely. My idea was to improve SettingsBundle a bit and allow each theme to define its own settings schema. Editable in the backend. This is something we need for a project as well so I totally see it coming relatively soon. It should not be that hard to implement, initial implementation could be as simple as:

```twig
{% set theme_settings = sylius_current_theme_settings() %}
```

The Twig helper would get the name of the current theme and load settings from appropriate namespace.

**Comment by pamil** [comment]
@steffenbrem settings are one of must-haves for themes, we should have them implemented in the near future. Anyway, I am curious, how did `settings.home_featured_products` become a taxon?

**Comment by aramalipoor** [comment]
We did a simple theme settings based on SyliusThemeBundle and SyliusSettingsBundle. It consists of a few services:

* **DynamicSettingsSchema** which implements `SchemaInterface` and works with service below,
* **DynamicSettingsBuilder** which accepts `FormBuilderInterface` and `SettingsBuilderInterface` and fills them up based a provided array definition of fields.
* **ThemeSettingsCompilerPass** a compiler pass which registers a settings schema for each theme based on fields defined in `theme.json` (using `DynamicSettingsSchema` class)

Anyway `DynamicSettingsBuilder` / `DynamicSettingsSchema` can be very useful services to have in SettingsBundle. All bundles can add array translators to a registry and we're gonna have an **array** to **Form** translator which is a useful thing in API era, and scenarios like theme settings.

This is a sample of our `theme.yml`s
```yml
settings:
  - id: header_taxonomy
    label: "Header Taxonomy"
    type: taxonomy
  - id: logo
    label: "Logo"
    type: image
  - id: footer_copyright
    label: "Footer Copyright"
    type: string
```

**Comment by steffenbrem** [comment]
@pamil I used proxy-like objects as global twig variables. They mostly implement `IteratorAggregate` or just use the magic `__call`, so you can lazy load settings, taxons, products etc. I liked the notation in twig to be like `settings.<setting_id>`. Then the type of the setting (`boolean`, `taxon` etc.) decides what it resolves to.

A pseudo proxy, that does this would look something like this:

```php
class SettingsProxy
{
    protected $valueResolver;

    public function __call($name, $arguments)
    {
        $setting = $this->findSettingByIdentifier($name);

        if (!$setting) {
            throw new \RuntimeException(sprintf(
                'Theme setting "%s" does not exist.',
                $name
            ));
        }

        // resolves a theme settings value to its proper form (might be a `Taxon` entity for example)
        return $this->valueResolver->resolve($setting);
    }
}
```

**Comment by aramalipoor** [comment]
@steffenbrem This is a great way ;)

**Comment by steffenbrem** [comment]
Why would you use the settings bundle for this, what is it's benefit? I am using a dedicated model/resource `Mango\Component\Theme\Model\Setting` to store theme settings. This setting has a property `configuration`, which is a serialized array, in which is stores for example a taxons id, or the value of a boolean setting.

I wonder how the implementation with settings bundle would differ from mine?

**Comment by pjedrzejewski** [comment]
@steffenbrem Settings bundle takes care of form editing and has many usecase and is already in Sylius. I basically would like to avoid introducing another system for something that we have solved. :) I like your solution and I think we could merge the ideas.

**Comment by pamil** [comment]
Theme settings support was implemented in #4873.
